Sync is causing out of memory errors

Greetings,

I am getting out of memory errors while trying to sync. Please see the following logcat around line 1108:
https://gist.github.com/jcamp1/68c666a959d8ca35f2c1

Thanks,
Justin

note to self:

Potential root cause

##### OutOfMemoryError was thrown from `Utils.decompressByGzip()`. 
```
W/System.err(15016): java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError
W/System.err(15016): 	at java.util.concurrent.FutureTask$Sync.innerGet(FutureTask.java:223)
W/System.err(15016): 	at java.util.concurrent.FutureTask.get(FutureTask.java:82)
W/System.err(15016): 	at com.couchbase.lite.replicator.PullerInternal.waitForPendingFutures(PullerInternal.java:932)
W/System.err(15016): 	at com.couchbase.lite.replicator.PullerInternal$14.run(PullerInternal.java:1002)
I/APACHE HTTP (thCr=1635) - NafHttpAuthStrategyDefault(15016): (thUse=1635) NafHttpAuthStrategyDefault()
W/System.err(15016): 	at java.lang.Thread.run(Thread.java:856)
W/System.err(15016): Caused by: java.lang.OutOfMemoryError
W/System.err(15016): 	at java.io.ByteArrayOutputStream.expand(ByteArrayOutputStream.java:91)
W/System.err(15016): 	at java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:201)
W/System.err(15016): 	at com.couchbase.lite.util.Utils.decompressByGzip(Utils.java:213)
W/System.err(15016): 	at com.couchbase.lite.support.MultipartDocumentReader.parseJsonBuffer(MultipartDocumentReader.java:45)
W/System.err(15016): 	at com.couchbase.lite.support.MultipartDocumentReader.finishedPart(MultipartDocumentReader.java:238)
W/System.err(15016): 	at com.couchbase.lite.support.MultipartReader.appendData(MultipartReader.java:171)
W/System.err(15016): 	at com.couchbase.lite.support.MultipartDocumentReader.appendData(MultipartDocumentReader.java:88)
W/System.err(15016): 	at com.couchbase.lite.replicator.BulkDownloader.appendToPart(BulkDownloader.java:236)
W/System.err(15016): 	at com.couchbase.lite.support.MultipartReader.trimBuffer(MultipartReader.java:122)
W/System.err(15016): 	at com.couchbase.lite.support.MultipartReader.appendData(MultipartReader.java:176)
W/System.err(15016): 	at com.couchbase.lite.replicator.BulkDownloader.executeRequest(BulkDownloader.java:158)
W/System.err(15016): 	at com.couchbase.lite.replicator.BulkDownloader.run(BulkDownloader.java:92)
W/System.err(15016): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:442)
W/System.err(15016): 	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
W/System.err(15016): 	at java.util.concurrent.FutureTask.run(FutureTask.java:137)
W/System.err(15016): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150)
W/System.err(15016): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264)
W/System.err(15016): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
W/System.err(15016): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
```

##### Number of docs or doc size could cause this issue.

##### ByteArrayInputStream and GZIPInputStream should be closed in finally block
```
public static byte[] decompressByGzip(byte[] sourceBytes){
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try {
            try {
                byte[] buffer = new byte[CHUNK_SIZE];
                ByteArrayInputStream in = new ByteArrayInputStream(sourceBytes);
                GZIPInputStream gzip = new GZIPInputStream(in);
                int len = 0;
                while ((len = gzip.read(buffer, 0, CHUNK_SIZE)) != -1) {
                    out.write(buffer, 0, len);
                }
            }
            catch (IOException ex){
                return null;
            }
            return out.toByteArray();
        }
        finally {
            try{ out.close(); }catch(IOException ex){}
        }
    }
```

--- a/src/main/java/com/couchbase/lite/replicator/BulkDownloader.java
+++ b/src/main/java/com/couchbase/lite/replicator/BulkDownloader.java
@@ -173,19 +173,17 @@ public class BulkDownloader extends RemoteRequest implements MultipartReaderDele
                 else {
                     Log.v(Log.TAG_SYNC, "contentTypeHeader is not multipart = %s",contentTypeHeader.getValue());
                     if (entity != null) {
-
-                        inputStream = entity.getContent();
-                        // decompress if contentEncoding is gzip
-                        if(Utils.isGzip(entity)){
-                            inputStream = new GZIPInputStream(inputStream);
-                        }
                         try {
-                            fullBody = Manager.getObjectMapper().readValue(inputStream,
-                                    Object.class);
+                            inputStream = entity.getContent();
+                            // decompress if contentEncoding is gzip
+                            if(Utils.isGzip(entity)){
+                                inputStream = new GZIPInputStream(inputStream);
+                            }
+                            fullBody = Manager.getObjectMapper().readValue(inputStream, Object.class);
                             respondWithResult(fullBody, error, response);
                         } finally {
                             try {
-                                inputStream.close();
+                                if(inputStream != null) { inputStream.close(); }
                             } catch (IOException e) {
                             }
                         }

--- a/src/main/java/com/couchbase/lite/support/RemoteMultipartDownloaderRequest.java
+++ b/src/main/java/com/couchbase/lite/support/RemoteMultipartDownloaderRequest.java
@@ -121,19 +121,18 @@ public class RemoteMultipartDownloaderRequest extends RemoteRequest {
                 }
                 else {
                     if (entity != null) {
-                        inputStream = entity.getContent();
-
-                        // decompress if contentEncoding is gzip
-                        Header contentEncoding = entity.getContentEncoding();
-                        if(contentEncoding != null && contentEncoding.getValue().contains("gzip")){
-                            inputStream = new GZIPInputStream(inputStream);
-                        }
                         try {
+                            inputStream = entity.getContent();
+                            // decompress if contentEncoding is gzip
+                            Header contentEncoding = entity.getContentEncoding();
+                            if(contentEncoding != null && contentEncoding.getValue().contains("gzip")){
+                                inputStream = new GZIPInputStream(inputStream);
+                            }
                             fullBody = Manager.getObjectMapper().readValue(inputStream, Object.class);
                             respondWithResult(fullBody, error, response);
                         } finally {
                             try {
-                                inputStream.close();
+                                if(inputStream != null){ inputStream.close(); }
                             } catch (IOException e) {
                             }
                         }

--- a/src/main/java/com/couchbase/lite/support/RemoteRequest.java
+++ b/src/main/java/com/couchbase/lite/support/RemoteRequest.java
@@ -222,7 +222,7 @@ public class RemoteRequest implements Runnable {
                         fullBody = Manager.getObjectMapper().readValue(inputStream, Object.class);
                     } finally {
                         try {
-                            inputStream.close();
+                            if(inputStream != null){ inputStream.close(); }
                         } catch (IOException e) {
                         }
                     }

--- a/src/main/java/com/couchbase/lite/util/Utils.java
+++ b/src/main/java/com/couchbase/lite/util/Utils.java
@@ -183,13 +183,17 @@ public class Utils {
     public static byte[] compressByGzip(byte[] sourceBytes){
         ByteArrayOutputStream out = new ByteArrayOutputStream();
         try{
+            GZIPOutputStream gzip = null;
             try {
-                GZIPOutputStream gzip = new GZIPOutputStream(out);
+                gzip = new GZIPOutputStream(out);
                 gzip.write(sourceBytes);
                 gzip.close();
             }
             catch (IOException ex){
                 return null;
+            }
+            finally {
+                try{ if(gzip != null) { gzip.close(); } } catch(IOException ex) {}
             }
 
             return  out.toByteArray();
@@ -202,24 +206,29 @@ public class Utils {
     // from gzip
     public static int CHUNK_SIZE = 8192; // 1024 * 8
     public static byte[] decompressByGzip(byte[] sourceBytes){
-        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        byte[] buffer = null;
+        ByteArrayOutputStream out = null;
+        ByteArrayInputStream in = null;
+        GZIPInputStream gzip = null;
         try {
-            try {
-                byte[] buffer = new byte[CHUNK_SIZE];
-                ByteArrayInputStream in = new ByteArrayInputStream(sourceBytes);
-                GZIPInputStream gzip = new GZIPInputStream(in);
-                int len = 0;
-                while ((len = gzip.read(buffer, 0, CHUNK_SIZE)) != -1) {
-                    out.write(buffer, 0, len);
-                }
-            }
-            catch (IOException ex){
-                return null;
+            out    = new ByteArrayOutputStream();
+            buffer = new byte[CHUNK_SIZE];
+            in     = new ByteArrayInputStream(sourceBytes);
+            gzip   = new GZIPInputStream(in);
+            int len = 0;
+            while ((len = gzip.read(buffer, 0, CHUNK_SIZE)) != -1) {
+                out.write(buffer, 0, len);
             }
             return out.toByteArray();
         }
+        catch (IOException ex){
+            Log.w(Log.TAG, "Failed to decompress gzipped data: " + ex.getMessage());
+            return null;
+        }
         finally {
-            try{ out.close(); }catch(IOException ex){}
+            try{ if(out  != null) { out.close();  } } catch(IOException ex) {}
+            try{ if(in   != null) { in.close();   } } catch(IOException ex) {}
+            try{ if(gzip != null) { gzip.close(); } } catch(IOException ex) {}
         }
     }
 

